python/rmm/_lib/lib.pxd

```diff
@@ -0,0 +1,162 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+from libc.stdint cimport uintptr_t
+from libcpp cimport bool
+from libcpp.utility cimport pair
+
+ctypedef long int offset_t
+ctypedef pair[const char*, unsigned int] caller_pair
+
+
+cdef extern from * nogil:
+
+    ctypedef void* cudaStream_t "cudaStream_t"
+
+
+cdef uintptr_t c_alloc(
+    size_t size,
+    cudaStream_t stream
+) except? <uintptr_t>NULL
+
+cdef uintptr_t c_realloc(
+    size_t new_size,
+    cudaStream_t stream
+) except? <uintptr_t>NULL
+
+cdef void c_free(
+    void *ptr,
+    cudaStream_t stream
+)
+
+cdef offset_t* c_getallocationoffset(
+    void *ptr,
+    cudaStream_t stream
+) except? <offset_t*>NULL
+
+cdef caller_pair _get_caller()
+
+
+cdef extern from "rmm/rmm.h" nogil:
+
+    ctypedef enum rmmError_t:
+        RMM_SUCCESS = 0,
+        RMM_ERROR_CUDA_ERROR,
+        RMM_ERROR_INVALID_ARGUMENT,
+        RMM_ERROR_NOT_INITIALIZED,
+        RMM_ERROR_OUT_OF_MEMORY,
+        RMM_ERROR_UNKNOWN,
+        RMM_ERROR_IO,
+        N_RMM_ERROR,
+
+    ctypedef enum rmmAllocationMode_t:
+        CudaDefaultAllocation = 0,
+        PoolAllocation = 1,
+        CudaManagedMemory = 2,
+
+    ctypedef struct rmmOptions_t:
+        rmmAllocationMode_t allocation_mode
+        size_t initial_pool_size
+        bool enable_logging
+
+    cdef rmmError_t rmmInitialize(
+        rmmOptions_t *options
+    ) except +
+
+    cdef rmmError_t rmmFinalize() except +
+
+    cdef bool rmmIsInitialized(
+        rmmOptions_t *options
+    ) except +
+
+    cdef const char * rmmGetErrorString(
+        rmmError_t errcode
+    ) except +
+
+    cdef rmmError_t rmmAlloc(
+        void **ptr,
+        size_t size,
+        cudaStream_t stream,
+        const char* file,
+        unsigned int line
+    ) except +
+
+    cdef rmmError_t rmmRealloc(
+        void **ptr,
+        size_t new_size,
+        cudaStream_t stream,
+        const char* file,
+        unsigned int line
+    ) except +
+
+    cdef rmmError_t rmmFree(
+        void *ptr,
+        cudaStream_t stream,
+        const char* file,
+        unsigned int line
+    ) except +
+
+    cdef rmmError_t rmmGetAllocationOffset(
+        offset_t *offset,
+        void *ptr,
+        cudaStream_t stream
+    ) except +
+
+    cdef rmmError_t rmmGetInfo(
+        size_t *freeSize,
+        size_t *totalSize,
+        cudaStream_t stream
+    ) except +
+
+    cdef rmmError_t rmmWriteLog(
+        const char* filename
+    ) except +
+
+    cdef size_t rmmLogSize() except +
+
+    cdef rmmError_t rmmGetLog(
+        char* buffer,
+        size_t buffer_size
+    ) except +
+
+
+cdef extern from "rmm/rmm.hpp" namespace "rmm" nogil:
```

Comment:
Note that I think eventually we may eliminate the C interface above and everything will move to the C++ API. CFFI was blocking this.

Reply:
Yup understood, will adjust the Cython in lockstep with those changes, didn't want to rock the boat for now though 😄

Reply:
Yup understood, will adjust the Cython in lockstep with those changes, didn't want to rock the boat for now though 😄